src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,106 @@
 package core.basesyntax;
 
+import java.util.Objects;
+
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+    private int size;
+    private int capacity;
+    private int threshold;
+    private Node<K, V>[] table;
+
+    public MyHashMap() {
+        capacity = DEFAULT_CAPACITY;
+        threshold = (int) (LOAD_FACTOR * capacity);
+        table = (Node<K, V>[]) new Node[capacity];
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (++size > threshold) {
+            resize();
+        }
+        setNode(new Node<>(getHash(key), key, value, null));
     }
 
     @Override
     public V getValue(K key) {
+        if (size == 0) {
+            return null;
+        }
+        Node<K, V> localNode = table[getHash(key) % capacity];
+        while (localNode != null) {
+            if (Objects.equals(localNode.key, key)) {
+                return localNode.value;
+            }
+            localNode = localNode.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int getHash(K key) {
+        return key == null ? 0 : Math.abs(key.hashCode());
+    }
+
+    private void resize() {
+        capacity = capacity << 1;
+        threshold = (int) (capacity * LOAD_FACTOR);
+        Node<K, V>[] localTable = table;
+        table = (Node<K, V>[]) new Node[capacity];
+        for (int i = 0; i < localTable.length; i++) {
+            while (localTable[i] != null) {
+                put(localTable[i].key, localTable[i].value);
+                localTable[i] = localTable[i].next;
+                size--;
+            }
+        }
+    }
+
+    private void setNode(Node<K, V> node) {
+        Node<K, V> localNode = table[node.hash % capacity];
+        if (localNode == null) {
+            table[node.hash % capacity] = node;
+            return;
+        }
+        while (localNode.next != null) {
+            if (checkNodes(node, localNode)) {
+                return;
+            }
+            localNode = localNode.next;
+        }
+        if (checkNodes(node, localNode)) {
+            return;
+        }
+        localNode.next = node;
+    }
+
+    private boolean checkNodes(Node<K,V> node1, Node<K,V> node2) {
+        if (Objects.equals(node1.key, node2.key)) {
+            size--;
+            node2.value = node1.value;
+            return true;
+        }
+        return false;
+    }
+
+    private static class Node<K, V> {
+
+        private final int hash;
```

Comment:
where do we use this field? let's remove it, it's redundant in this implementation

Reply:
done